odoo/tools/translate.py

```diff
@@ -476,257 +477,279 @@ def unquote(str):
-        self.buffer = codecs.StreamReaderWriter(
-            stream=buffer,
-            Reader=codecs.getreader('utf-8'),
-            Writer=codecs.getwriter('utf-8')
-        )
+def TranslationFileReader(source, fileformat='po'):
+    """ Iterate over translation file to return Odoo translation entries """
+    if fileformat == 'csv':
+        return CSVFileReader(source)
+    if fileformat == 'po':
+        return PoFileReader(source)
+    _logger.info('Bad file format: %s', fileformat)
+    raise Exception(_('Bad file format: %s') % fileformat)
+
+class CSVFileReader:
+    def __init__(self, source):
+        self.source = pycompat.csv_reader(source, quotechar='"', delimiter=',')
+        # read the first line of the file (it contains columns titles)
+        self.fields = next(self.source)
 
     def __iter__(self):
-        self.buffer.seek(0)
-        self.lines = self._get_lines()
-        self.lines_count = len(self.lines)
-
-        self.first = True
-        self.extra_lines= []
-        return self
-
-    def _get_lines(self):
-        lines = self.buffer.readlines()
-        # remove the BOM (Byte Order Mark):
-        if len(lines):
-            lines[0] = lines[0].lstrip(u"\ufeff")
-
-        lines.append('') # ensure that the file ends with at least an empty line
-        return lines
-
-    def cur_line(self):
-        return self.lines_count - len(self.lines)
-
-    def next(self):
-        trans_type = name = res_id = source = trad = module = None
-        if self.extra_lines:
-            trans_type, name, res_id, source, trad, comments = self.extra_lines.pop(0)
-            if not res_id:
-                res_id = '0'
+        for entry in self.source:
+            yield zip(self.fields, entry)
+
+class PoFileReader:
+    """ Iterate over po file to return Odoo translation entries """
+    def __init__(self, source):
+
+        def get_pot_path(source_name):
+            # when fileobj is a TemporaryFile, its name is an interget in P3, a string in P2
+            if isinstance(source_name, str) and source_name.endswith('.po'):
+                # Normally the path looks like /path/to/xxx/i18n/lang.po
+                # and we try to find the corresponding
+                # /path/to/xxx/i18n/xxx.pot file.
+                # (Sometimes we have 'i18n_extra' instead of just 'i18n')
+                addons_module_i18n, _ignored = os.path.split(source_name)
+                addons_module, i18n_dir = os.path.split(addons_module_i18n)
+                addons, module = os.path.split(addons_module)
+                pot_path = os.path.join(
+                    addons, module, i18n_dir, module + '.pot')
+                return os.path.exists(pot_path) and pot_path or False
+            return False
+
+        # polib accepts either a path, either the file content as a string
+        # NamedTemporaryFile is not a BufferedIOBase but a wrapper
+        if isinstance(source, io.BufferedIOBase) or \
```

Comment:
Why not go with the second branch if source is a string, and the first otherwise? If `source` is a `BufferedIOBase` without an actual backing file it's going to go through the second path and blow up anyway.